rfcs/convergence/root-as-slot.md

```diff
@@ -0,0 +1,248 @@
+# RFC: root as a slot
+
+---
+
+_@bsunderhus_
+
+## Summary
+
+This RFC proposes to treat `root` as a _slot_ to simplify signature, improve `Typings` and avoid conflicts between `root` and other _slots_ signatures.
+
+Main changes would be:
+
+1. `root` becomes a regular slot on component state
+2. Support references on every _slot_
+3. Supports Custom Elements on every _slot_
+4. Allows a restricted `as` prop on every slot
+5. stops using `getNativeElementProps` on every slot
+
+## Problem statement
+
+There are disparities in how `root` is treated compared to _slots_. Some discussions have popped up due to these divergences between _slots_ and `root`:
+
+1. [Do not support as prop for slots that render native DOM elements] conflicts with [1st rule of ARIA opt-out mechanism]
+2. Support for Custom Elements on _slots_ but not for _root_
+3. Right now `Typings` for `root` aren't available as there's no way to split `root` from internal state.
+4. There's no possible way of passing refs to slots but `root` does
+5. [Primary slot]
+
+Major differences between `root` 🌿 and _slots_ 🎰:
+
+1. References 🌿
+2. Custom Elements as base element 🎰
+3. `as` prop 🌿
+4. Well defined Types 🎰
+
+## Detailed Design or Proposal
+
+For the User of the component perspective there will be no changes in the signature.
+
+### Converge `root` to a _slot_
+
+This RFC proposes to treat `root` as a _slot_.
+
+Treating `root` as a slot means having to declare `root` _Typings_ together with other _slots_. This will impact in a components state signature, where `root` will have to be declared.
+
+> ⚠️ That doesn't mean `root` will be available in component's properties interface
+
+What we have right now:
+
+```ts
+export type Slots = {
+  slot: ObjectShorthandProps<SlotProps>;
+};
+export interface Props extends ComponentProps<Slots> & React.HTMLAttributes<HTMLElement> {/* ... */}
+export interface State extends ComponentState<Slots> {/* ... */}
+
+// use*State hook
+function useState({ slot, ...props }: Props): State {
+  // ...
+  return {
+    ...props,
+    slot: resolveShorthand(slot),
+  };
+}
+```
+
+With the modifications:
+
+```ts
+export type Slots = {
+  root: ObjectShorthandProps<React.HTMLAttributes<HTMLElement>, HTMLElement>;
+  slot: ObjectShorthandProps<SlotProps>;
+};
+export interface Props extends ComponentProps<Slots> {
+  /* ... */
+}
+export interface State extends ComponentState<Slots> {
+  /* ... */
+}
+
+// use*State hook
+function useState({slot, ...props}: Props): State {
+  // ...
+  return {
+    slot: resolveShorthand(slot)
+    root: resolveShorthand(props),
```

Comment:
```suggestion
    root: resolveRoot(props),
```

or

```suggestion
    root: getNativeElementProps(props),
```


As we discussed in offline, we cannot use `resolveShorthand()` for slot as it accepts component's props.

---

We need to have `props` filtering to avoid leaking props, for example:

```tsx
<SampleComponent size="large" sampleSlot={{ role: 'none' }} />
// 🔽
const state = { root: resolveShorthand({ size: 'large', sampleSlot: { role: 'none' } }) }
// state.root will contain `sampleSlot` & `size` 💣
// they will be spreaded on DOM
```

---

Overall, there is no sense to call `resolveShorthand()` on `props` as `props` are already an object 🤨 `resolveRoot()` might be a better solution as we need to return specific interface:

```tsx
resolveShorthand(props /* ComponentProps */) // returns "ObjectShorthandProps<??>
```

Reply:
I don't know if `resolveRoot` is a good name, perhaps the best idea would be to explicitly use the native props filtering, but still, this is debatable